pkg/bus/bus.go

```diff
@@ -0,0 +1,60 @@
+package bus
+
+import (
+	"sync"
+
+	"github.com/infrawatch/sg-core-refactor/pkg/data"
+)
+
+/* TODO: optimize this
+channels are not particularily fast, best to avoid them in the fastpath and use Mutexes instead
+benchmark implementation with callback vs channels here
+*/
+
+//EventBus bus for data.Event type
+type EventBus struct {
+	subscribers []chan data.Event
+	rw          sync.RWMutex
+}
+
+//Subscribe subscribe to bus
+func (eb *EventBus) Subscribe(c chan data.Event) {
+	eb.rw.Lock()
+	defer eb.rw.Unlock()
+	eb.subscribers = append(eb.subscribers, c)
+}
+
+//Publish publish to bus
+func (eb *EventBus) Publish(e data.Event) {
+	eb.rw.RLock()
+	for _, c := range eb.subscribers {
+		go func(c chan data.Event, e data.Event) {
+			c <- e
+		}(c, e)
+	}
+	eb.rw.RUnlock() //defer is actually very slow
```

Comment:
I believe the go func at line 32 means that you won't block, but you could possibly see a crash if you run out of go routine space.  With the sg-bridge, overflows are treated as part of the communication channel (i.e. lost udp packet). For metrics, discarding is fine as there should be another metric coming along soon enough.  The sg-bridge uses a ring buffer so that the oldest metric is thrown away.  Can't really do that here.  What about using a non-blocking send?  i.e.

```
for _, c := range eb.subscribers {
  select {
      case c <- e:
        sent++
      default:
         dropped++
      }
}

The channels should be buffered to allow for goroutine scheduling.
```

Reply:
Same for metrics below...